csharp/src/Ice/Endpoint.cs

```diff
@@ -19,94 +19,170 @@ public interface Endpoint
 /// <summary>
 /// Base class providing access to the endpoint details.
 /// </summary>
-public abstract class EndpointInfo
+public class EndpointInfo
 {
     /// <summary>
-    /// The information of the underlying endpoint or null if there's no underlying endpoint.
+    /// Gets the information for the underlying endpoint, or null if there's no underlying endpoint.
     /// </summary>
-    public EndpointInfo? underlying;
+    public readonly EndpointInfo? underlying;
 
     /// <summary>
-    /// The timeout for the endpoint in milliseconds. -1 means no timeout.
+    /// Gets the timeout of the endpoint in milliseconds. -1 means no timeout.
     /// </summary>
-    public int timeout;
+    public readonly int timeout;
 
     /// <summary>
-    /// Specifies whether or not compression should be used if available when using this endpoint.
+    /// Gets a value indicating whether or not compression should be used if available when using this endpoint.
     /// </summary>
-    public bool compress;
+    public readonly bool compress;
 
-    /// <summary>Returns the type of the endpoint.</summary>
+    /// <summary>Gets a 16-bit integer that identifies the transport of this endpoint.</summary>
     /// <returns>The endpoint type.</returns>
-    public abstract short type();
-
-    /// <summary>Returns true if this endpoint is a datagram endpoint.</summary>
-    /// <returns>True for a datagram endpoint.</returns>
-    public abstract bool datagram();
-
-    /// <summary>Returns true if this endpoint is secure; otherwise false.</summary>
-    /// <returns>True if the endpoint is secure.</returns>
-    public abstract bool secure();
+    /// <remarks>The type of an underlying endpoint is always the same as the type its enclosing endpoint.</remarks>
+    public virtual short type() => underlying?.type() ?? -1;
```

Comment:
Should we instead assert underlying is not null?

Reply:
Even with a Debug.Assert, we would still need to return something. So I think -1 is fine here.